### PR TITLE
[POAE7-2911] Cider Parquet Reader to Arrow

### PR DESCRIPTION
**Pull request by imajiayu**
### What changes were proposed in this pull request?
new Class `CiderParquetReader::Reader`, which reads local parquet file to `ArrowSchema`&`ArrowArray` directly.

### How was this patch tested?
CiderParquetReaderTest.cpp

### CODE REVIEW

File: cpp/src/cider/tests/utils/CiderParquetReaderTest.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                                        actual_array,
+                                        std::get<0>(schema_and_array),
+                                        actual_schema));
+}
+
+// ┌──────────┐
+// │ col_type │
+// ├──────────┤
+// │ 0        │
+// │          │
```
Comment (jikunshang) [inline cpp/src/cider/tests/utils/CiderParquetReaderTest.cpp:66]: add null

File: cpp/src/cider/util/ArrowArrayBuilder.h [hunk 2]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright(c) 2022-2023 Intel Corporation.
+ * Copyright (c) OmniSci, Inc. and its affiliates.
```
Comment (jikunshang) [inline cpp/src/cider/util/ArrowArrayBuilder.h:3]: can remove?

Comment (imajiayu) [reply cpp/src/cider/util/ArrowArrayBuilder.h:3]: license-check will fail.

**Comment by jikunshang** [comment]
@xieqi Please take a review, thanks.

File: cpp/src/cider/tests/utils/CiderParquetReaderTest.cpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+// ┌──────────┐
+// │ col_type │
+// ├──────────┤
+// │ 0        │
+// │ null     │
+// │ 2        │
+// │ null     │
+// │ 4        │
+// └──────────┘
+#define GENERATE_READER_TEST(type_name, substrait_type, c_type)               \
```
Comment (jikunshang) [inline cpp/src/cider/tests/utils/CiderParquetReaderTest.cpp:74]: maybe we can use single file and read different columns?

**Review by jikunshang — DISMISSED** [review]

File: cpp/src/cider/util/CiderParquetReader.cpp [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+  currentBatchSize_ = batchSize;
+
+  int rowsActualRead = doReadBatch(rowsToRead, buffersPtr, nullsPtr);
+
+  totalRowsRead_ += rowsActualRead;
+  LOG(INFO) << "total rows read yet: " << totalRowsRead_;
+  LOG(INFO) << "ret rows " << rowsActualRead;
+
+  auto schema_and_array = convert2Arrow(rowsToRead, buffersPtr, nullsPtr);
```
Comment (xieqi) [inline cpp/src/cider/util/CiderParquetReader.cpp:158]: if read to arrow why not use parquet arrow api to read the column to chunkarry then convert to arrowarray by arrow bridge

File: cpp/src/cider/util/CiderParquetReader.cpp [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+namespace CiderParquetReader {
+Reader::Reader() {}
+Reader::~Reader() {}
+
+void Reader::init(std::string fileName,
+                  std::vector<std::string> requiredColumnNames,
+                  int firstRowGroup,
+                  int rowGroupToRead) {
+  file_ = arrow::io::ReadableFile::Open(fileName).ValueOrDie();
```
[the remark this answers is not in the document]
Comment (imajiayu) [reply cpp/src/cider/util/CiderParquetReader.cpp:34]: using the `ARROW_ASSIGN_OR_RAISE` need to change the return type from `void` to `arrow::status`. I add some codes to raise exception if open failed.

File: cpp/src/cider/util/CiderParquetReader.cpp (hunk 4 above)
Comment (jikunshang) [reply cpp/src/cider/util/CiderParquetReader.cpp:160]: we need to call `arrow::Table::CombineChunksToBatch()` first to convert to `RecordBatch`, and then call `ExportRecordBatch()`. the first step may have extra memory copy.

**Review by xieqi — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Pull request merged**